### DISCUSSION TITLE
What's your method to update values from Sliders to a byte value stored in Memoryblock?

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#269] [Q&A] [answered]
Hi everybody, 
I was wondering what's your usual method to update values in a sysex string from the different modulators on the panel?

This is the actual goal for my Yamaha SPX990 : 
I have stored a memoryblock of the sysex dump of the current preset [DONE]
The preset is decoded and all the parameter values reflects on the Panel [DONE]
When a modulator updates a value I need to send the CC value to the corresponding parameter to the SPX [DONE]
When a modulator updates a value I need to update the corresponding byte in the sysex dump stored in its memoryblock [REQUIRED]

Let's say for the parameter "Reverb Time" of the algo "reverb" I have the following modulator values registered:

`MainFxParamValue03 = SysexDecodedMessage:getByte(19)
`
then : 
```
	MainFxParam03 = panel:getModulatorByName("MainFxParam03")
	MainFxParam03:getComponent():setProperty("componentVisibility", 1, false)
	MainFxParam03:getComponent():setProperty("componentVisibleName", "RevTime", true)
	MainFxParam03:getComponent():setProperty("uiSliderMin", 0, true)
	MainFxParam03:getComponent():setProperty("uiSliderMax", 96, true)
	MainFxParam03:getComponent():setProperty("uiSliderDoubleClickEnabled", 1, true)
	MainFxParam03:getComponent():setProperty("uiSliderDoubleClickValue", 0, true)
	MainFxParam03:getComponent():setProperty("uiFixedSliderContent","0.3\n0.4\n0.5\n0.6\n0.7\n0.8\n0.9\n1.0\n1.1\n1.2\n1.3\n1.4\n1.5\n1.6\n1.7\n1.8\n1.9\n2.0\n2.1\n2.2\n2.3\n2.4\n2.5\n2.6\n2.7\n2.8\n2.9\n3.0\n3.1\n3.2\n3.3\n3.4\n3.5\n3.6\n3.7\n3.8\n3.9\n4.0\n4.1\n4.2\n4.3\n4.4\n4.5\n4.6\n4.7\n4.8\n4.9\n5.0\n5.5\n6.0\n6.5\n7.0\n7.5\n8.0\n8.5\n9.0\n9.5\n10.0\n11.0\n12.0\n13.0\n14.0\n15.0\n16.0\n17.0\n18.0\n19.0\n20.0\n25.0\n30.0\n35.0\n40.0\n45.0\n50.0\n55.0\n60.0\n65.0\n70.0\n75.0\n80.0\n85.0\n90.0\n95.0\n100.0\n120.0\n150.0\n180.0\n210.0\n240.0\n270.0\n300.0\n330.0\n360.0\n390.0\n420.0\n450.0\n480.0", true)
	MainFxParam03:getComponent():setValue(MainFxParamValue03-3, true) -- Starts at 3 for 0.3
	MainFxParam03:setProperty("modulatorCustomIndex","03", true)
	MainFxParam03:setProperty("modulatorCustomName","", true)
	MainFxParam03:setProperty("modulatorCustomIndexGroup","", true)
	MainFxParam03:setProperty("modulatorCustomNameGroup","", true)
	MainFxParam03:getComponent():setProperty("uiType","uiFixedSlider", true)
```

I need several datas and somewhere to store it : 
1. The byte position to modify  [OK]
2. The actual current Value [OK]
3. The calculation method to provide the compliant sysex value (direct from value, MSB /LSB for big values, negative values MSB/LSB with neg) [REQUIRED]
4. Then use copyFrom() to update the value to the memoryblock [OK]

My solution : 
1. The byte position could be set from "modulatorCustomIndexGroup"
2. The value is easy to get with getValue
3. The calculation method (should I use modulatorValueExpression or modulatorValueExpressionReverse or modulatorControllerExpression???  This one is important and I never worked with that.

Each modulator has its own calculation method. And depending on the current algo, the calculation method must be updated via mod/comp parameter setProperty etc

Most of the time the value to update to the memoryblock is the same as the modulator current value. 
But sometimes I need to add up the value to something like +5 or +6 etc
Sometimes I need to provide 2x bytes to the memoryblock for MSB[256] + LSB[256]
Sometimes I need to provide values MSB + LSB for "negative values" 2s complement

Here are some exemples of calculation method from the sysex message value to the modValue : 

1. DIRECT VALUE
`MainFxParamValue03 = SysexDecodedMessage:getByte(19)`

2. MSB LSB
`MainFxParam09RevDly = (MainFxParamValue12*256 + MainFxParamValue13)`

3. MSB LSB for NEGATIVE VALUES
```
	if 127-MainFxParamValue40 < 0 then  -- if MSB is between 127 and 255
		MainFxParam14TrgDlyMSB = (MainFxParamValue40*256)
		MainFxParam14TrgDlyMSB = 0 - (65536 - MainFxParam14TrgDlyMSB)
		MainFxParam14TrgDly = (MainFxParam14TrgDlyMSB + MainFxParamValue41)
		console("Displays MainFxParam14TrgDly : "..(MainFxParam14TrgDly)) -- MainFxParam14TrgDly  -100.0 to -1
	else
		MainFxParam14TrgDly = (MainFxParamValue40*256 + MainFxParamValue41)
		console("Displays MainFxParam14TrgDly : "..(MainFxParam14TrgDly)) -- MainFxParam14TrgDly   0 to +100.0
	end
```

Now I need the process on the other way : modValue to byte value.
I don't know exactly how to deal with that, it's not that complicated but I want to avoid something heavyweighted if I can go with something light as a feather, hence my questions.

Once I have the updated parameter value at the right byte position I will be able to send the sysex to the machine and we are good to go.

Thanks for your insight, I'm pretty sure you already dealt with something similar for synths that require updating parameters to the machine via sysex memory dumps.

Have a nice week to everyone

Damien

**Comment by Godlike-Productions** [comment, marked as answer]
You can use setByte (byteNumber,Value). Can't remember if you need a boolean at the end, but you may need to send true as well.

**Reply by dobo365** [reply]
Exactly.
I'm doing that all the time in my panels.
- documented sysex structure in one dummy method
- one shared memory block
- LoadData  using for example 	modMainVolume:setValue(LoadedProgramData:getByte(42), true)
- SaveData using for example 	LoadedProgramData:setByte(42, modMainVolume:getModulatorValue()) 

Those are then used by classical Load_OnChange, Save_OnChange methods attached to buttons

Good luck!

**Reply by damiensellier** [reply]
Thanks @dobo365 and @Godlike-Productions  for your replies. 
It's a bit more complicated than just storing the modulator value as it is in the MB. Before I have to translate them as mentioned before.
I was mostly wondering where do you store your destination value calculation methods, and your destination bytes position for each values?
Do you have multiple "global" value calculation methods (for Positive values, for LUT values, for negative values, MSB/LSB etc) and call the method via component property (1,2,3,4 ..n method stored as a flag in i.e as modulatorCustomIndexGroup) or you put the whole method of calculation every time for every parameter in the mod property "modulatorValueExpression" for exemple?  
Do you call a different method for each slider when the value change or you use a global method (I would go for a global since I have 60 sliders whose behaviour changes for 40 algorithm)?

Thanks
Damien

**Reply by Godlike-Productions** [reply]
Yeah, I took it one step further with the class structure.  The dummy method essentially becomes your prototype, and it gets wrapped up in a single data structure.  You can then just invoke a new class member with the same structure without having to redeclare everything.

**Reply by Godlike-Productions** [reply]
It very much depends on what I'm doing with them.  The last panel I wrote each memoryBlock stored a partial SYSEX dump (Data without the sysex headers).  I had a memory block as a table element along with name, author and category.  Then I had a table of those for each patch number.

I had a method that scanned the state of every modulator and wrote it into the memoryblock (Well a table first that I converted to a memory block).

When I recalled the memory block, I had a method that just read each byte and set each modulator.

They are long methods, but fairly simple.

I did them all by Modulator Name

**Reply by damiensellier** [reply]
@Godlike-Productions Ok I get it. So the big part of the job relies on creating tables that stores the destination byte position and so on for each parameter beforhand then call it when required based on the modulator name and probably other statements.  I thought about that but I think I can store some values in the component custom fields it would skip some steps. It seems unorthodox though :) I must admit I am fed up about this panel, I already wrote somthing like 800 mod (40 algos, 20to34 sliders each) with all the usual properties (10 :  min, max etc)  So I want to keep this part as simple as it could. Next time I'll work with 3D tables/arrays  with x: modulator/parameters, y:properties, z:Algos(reverb, chorus etc) That will remove most of the hard coded lines.

**Reply by dobo365** [reply]
Same for me. 
Two long methods (LoadProgramData and SaveProgramData) but just a big list of commands I stated above.
I keep things in the same order at different places, so you can get some help with an editor and search/replace but it is mainly a manual job and this is a part of the time spent on building a panel.

**Reply by damiensellier** [reply]
Let's go! Wish me good luck and everything should be fine today or tomorrow night. thanks ;)

**Reply by damiensellier** [reply]
@dobo365 @Godlike-Productions Thank you so much ! 
I did a mix of both of you recommanded and what I had in mind about the way to store the byte position to update and the value calculation method in the mod properties .

This is all the parameters used for a modulator : 

```
	MainFxParam04 = panel:getModulatorByName("MainFxParam04")
	MainFxParam04:getComponent():setProperty("componentVisibility", 1, false)
	MainFxParam04:getComponent():setProperty("componentVisibleName", "RevTime", true)
	MainFxParam04:getComponent():setProperty("uiSliderMin", 0, true)
	MainFxParam04:getComponent():setProperty("uiSliderMax", 96, true)
	MainFxParam04:getComponent():setProperty("uiSliderDoubleClickEnabled", 1, true)
	MainFxParam04:getComponent():setProperty("uiSliderDoubleClickValue", 0, true)
	MainFxParam04:getComponent():setProperty("uiFixedSliderContent","0.3\n0.4\n0.5\n0.6\n0.7\n0.8\n0.9\n1.0\n1.1\n1.2\n1.3\n1.4\n1.5\n1.6\n1.7\n1.8\n1.9\n2.0\n2.1\n2.2\n2.3\n2.4\n2.5\n2.6\n2.7\n2.8\n2.9\n3.0\n3.1\n3.2\n3.3\n3.4\n3.5\n3.6\n3.7\n3.8\n3.9\n4.0\n4.1\n4.2\n4.3\n4.4\n4.5\n4.6\n4.7\n4.8\n4.9\n5.0\n5.5\n6.0\n6.5\n7.0\n7.5\n8.0\n8.5\n9.0\n9.5\n10.0\n11.0\n12.0\n13.0\n14.0\n15.0\n16.0\n17.0\n18.0\n19.0\n20.0\n25.0\n30.0\n35.0\n40.0\n45.0\n50.0\n55.0\n60.0\n65.0\n70.0\n75.0\n80.0\n85.0\n90.0\n95.0\n100.0\n120.0\n150.0\n180.0\n210.0\n240.0\n270.0\n300.0\n330.0\n360.0\n390.0\n420.0\n450.0\n480.0", true)
	MainFxParam04:getComponent():setValue(MainFxParamValue03-3, true) -- Starts at 3 for 0.3
	MainFxParam04:setProperty("modulatorCustomIndex","03", true)
	MainFxParam04:setProperty("modulatorCustomName","3", true) -- Used to store the byte position to update
	MainFxParam04:setProperty("modulatorCustomIndexGroup","3", true) -- Used to store the cstnt to substrct to the modValue
	MainFxParam04:setProperty("modulatorCustomNameGroup","1", true) -- Used to store the calculation method (1,2,3 or 4)
	MainFxParam04:getComponent():setProperty("uiType","uiFixedSlider", true)

```

I have everything I need stored in the mod properties this way and it's a success. 
I have very short global methods for value updates and it's working like a charm without creating new arrays.
It was a big pain because the panel is based only on full memory dump that I had to decypher byte by byte (no documentation) and there's only 2 sysex functions : memory dump request/send, program change and a CC for a modulation pedal input to 1 parameter . This was extremely limited. But Now playing pingpong between the panel and the machine with those messages is enough to control every single parameters.


Thank you so much for the help
